README.md
# MQ2FarmTest

A MacroQuest plugin


changes.txt
2022-04-18: Chatwiththisname
	- Made an adjustment to account for bards casting.
	- Rearranged changes.txt so that the notes were more uniformly placed in the file (previously
		was oldest top, newest bottom and dates were a different format. )

2021-09-16: Chatwiththisname
	- Added consideration for MQ2Headshot which would cause it not to pull due to the (LastNameHere) under spawn names.
	- Changed the way the plugin checks for a spell to cast.
	- Updated how we determine what our current PercentHealth/Mana/Endurance is.

2021-09-15: Knightly
	- Conversion to Next

2019-01-29: Chatwiththisname
	- CHANGE: Changed the distance checks to Distance3DToSpawn function instead of GetDistance
    - CHANGE: No longer checking Health/Mana/Endurance for you or group while you have Aggro.
	- CHANGE: Adjust a "radius" circle on MQ2Map to depict the Radius engage
	- CHANGE: Added /farm command param "ShowSettings" to display all valid settings.
	- CHANGE: Gave radius a purpose. Set it so that where you issue the /farm on command is the anchor point for all potential
		target consideration, and the user should not select entities outside of that circle when farming.
	- CHANGE: Added the /permignore command to replicate the usage in the macro
	- CHANGE: Added DeathChecks and Missing Group members checks. These are the same thing as far as the plugin is concerned.
	- CHANGE: DeathCheck includes mercenary's correctly. Not yet handling reviving them.

2019-01-28: Chatwiththisname
	- CHANGE: If the pSpawn has no ManaMax, lie and say they're at 100% mana, even if they don't have a mana bar (rogue merc fix)
	- CHANGE: Added a delay after a disc is used before sitting again to avoid interupting discs during downtimes
	- CHANGE: Tell user about the "/farm help" command on initilize.
	- CHANGE: "/farm off" doesn't turn off attack if it's on? Bug or feature? Check for aggro, if none, turn off
		attack, otherwise leave it on?
	- CHANGE: Add a message to let the user know there are no valid targets nearby and you are sitting while waiting for more.

2019-01-22: Chatwiththisname
	- CHANGE: Fixed a typo in setting the ignore file, changed from FarmMobIgnores.ini to FarmMobIgnored.ini

2019-01-20: Chatwiththisname
	- Added the use of the FarmMobIgnored.ini file in the macro folder for the global ignores
		so players didn't need to reinvent the wheel and could use the existing ignores file.
	- Added group health mana and endurance checks so you can now have the merc tag along
		without worry of running them out of mana.
	- Reduced the delay between pulses by 1/3rd
	- Nav will now refresh destination if target is moving while navigating to the selected spawn.
	- Made some adjustments to stopping distance to prevent "You are to far away" issues.
	- Discs now verify target is in range before attempting to use if you have a target.
	- Fixed a crash that occured if no valid spawns were found.

2019-01-07: Chatwiththisname
	- Added more refined sorting for summmoning discs and Endurance Regenaration Discs.
	- Adding abilities base on the INI is no longer Rank Specific.
	- Removing Combat Abilities base on the INI is no longer rank specific.
	- INI Name now updated when the farm command is used for anything to ensure that the proper file is being written to.
	- Added in the casting of Aura Discs and Endurance Regen Discs for use out of combat.
	- Refined ActiveDisc detections to eliminate trying to use a disc when there is already an active disc, but allow the use of combat abilities unhindered.

2019-01-06: Chatwiththisname
	- INI's are now character specific, stored in release folder as MQ2FarmTest_CharacterName.ini
	- Discs are now grabbed and sorted when the /farm on command is given instead of when the plugin loads.

2019-01-05: Chatwiththisname
	- Got the Farm datatype fixed and added two members. TargetID and Version. ${Farm.TargetID} ${Farm.Version}
	- Cleaned up FarmCommand and it now report invalid command.
	- Made additions to ignorethese and ignorethis.
	- "/farm " Help list of commands updated to show all commands for current build.
	- Now sitting when no target is found to kill while it waits for more things to spawn or wander in range.
	- DiscReady function now makes sure you aren't trying to use an active disc when you already have one running.
	- Added usage of Instant cast beneficial buffs while in combat.
	- Made a HUD for use with the ${Farm.TargetID} to get a display of what it's going after and some information, useful for debugging. :-)
	- You can now add and remove Discs with INI entries. [DiscRemove] section for removing. [DiscAdd] section for Additions.

2019-01-03: Chatwiththisname
	- Now generating a list of Discs for toons to use and separating them to sections.
	- Firing Discs during combat based on the previously mentioned list.
	- DiscReady function now checks endurance/mana/reagent requirements before reporting true.
	- Actually enjoying testing this on my berserker. He's doing pretty good DPS without MQ2Melee.

2019-01-02: Chatwiththisname
	- Cleaning up the code to remove all the comments.
	- trying to add Discs to combat routines, not having much luck.
	- Separated FarmTest into seperate files.
	- Tried adding TLO and member support. No luck, but kept the file.

2018-11-12: Chatwiththisname
	- Working on CastDetrimental() function to cast single target detrimental spells.
	- Added rest routines.

2018-10-16: Chatwiththisname
	- Test has been down a while, but added a getFirstAggroed() function to try and
	- deal with aggro before trying to find a mob with a shortest path to fix
	- ignoring aggrod mobs.

2018-10-08: Chatwiththisname
	- Working on Algorithms for Farming.
	- Adding FarmCommand Functionality.
	- Added Health/Endurance/Mana Checks, still requires logic for resting routines.
	- Added HaveAggro() function. Returns true if any Autohaters are populated on Xtargetlist.

2018-10-07: Chatwiththisname/Renji
	- Worked SpawnSearch and Nav Code


FarmType.h
class MQ2FarmType : public MQ2Type
{
public:
	enum FarmMembers
	{
		TargetID = 1,
		Version = 2,
	};

	MQ2FarmType() :MQ2Type("Farm")
	{
		TypeMember(TargetID);//1,
		TypeMember(Version);//2,
	}

	virtual bool GetMember(MQVarPtr VarPtr, const char* Member, char* Index, MQTypeVar& Dest) override
	{
		MQTypeMember* pMember = FindMember(Member);
		if (!pMember)
			return false;
		switch (pMember->ID)
		{
		case TargetID:
			Dest.DWord = MyTargetID;
			Dest.Type = datatypes::pIntType;
			return true;
		case Version:
			sprintf_s(DataTypeTemp, "%1.2f", MQ2Version);
			Dest.Ptr  = &DataTypeTemp[0];
			Dest.Type = datatypes::pStringType;
			return true;
		}
		return false;
	}
};

inline class MQ2FarmType *pFarmType = nullptr;

inline bool dataFarm(const char* szIndex, MQTypeVar& Ret)
{
	if (szIndex != NULL)
	{
		//${Farm[szIndex]} gets used, the output goes here.
	}

	Ret.DWord = 1;
	Ret.Type = pFarmType;
	return true;
}

Variables.h
//variables
bool activated = false;
bool CastDetrimental = true;
bool Debugging = false;
bool DeadGroupMember = false;
bool GettingEndurance = false;
bool GettingHealth = false;
bool GettingMana = false;
bool useLogOut = false;
bool useEQBC = false;
bool useMerc = false;
bool useDiscs = true;
bool PullRequiresLineOfSight = true;
bool Paused = false;


char IgnoreINI[MAX_STRING] = { 0 };
char FarmMob[MAX_STRING] = { 0 };
char PullAbility[MAX_STRING] = { 0 };
char searchString[MAX_STRING] = { 0 };
char ThisINIFileName[MAX_STRING] = "";
char IgnoresFileName[MAX_STRING] = "";


unsigned long MyTargetID = 0;

float AnchorX = 0;
float AnchorY = 0;
float AnchorZ = 0;

uint64_t CastLastTimeUsed = GetTickCount64();
uint64_t DiscLastTimeUsed = GetTickCount64();
int HealAt = 0;
int HealTill = 0;
int MedEndAt = 0;
int MedEndTill = 0;
int MedAt;
int MedTill = 0;

int CampRadius = 0;
int GemIndex = 0;
int PullAbilityRange = 0;
int PulseDelay = 20;
int Pulse = 0;
int Radius = 0;
int ZRadius = 0;

std::vector<PSPELL> SingleDetrimental; //used in combat
std::vector<PSPELL> AEDetrimental; //used in combat
std::vector<PSPELL> ToTBeneficial; //used in combat
std::vector<PSPELL> Summons; //summons items
std::vector<PSPELL> EndRegen; //Endurance Regen Discs
std::vector<PSPELL> SingleBeneficial; //out of combat
std::vector<PSPELL> SelfBeneficial; //out of combat
std::vector<PSPELL> GroupBeneficial; //out of combat
std::vector<PSPELL> Aura;

//End Variables

Prototypes.h
//Prototypes
bool AmIReady();
bool atob(char* x);
bool DiscReady(PSPELL);
bool HaveAggro();
bool IHaveBuff(PSPELL);
bool MeshLoaded();
bool NavActive();
bool NavPaused();
bool PathExists(unsigned long SpawnID);
bool SpellsMemorized();

unsigned long getFirstAggroed();
unsigned long SearchSpawns(char* szIndex);

float AmFacing(unsigned long ID);
float PathLength(unsigned long SpawnID);

inline bool Casting();
inline bool InGame();
inline float PercentMana(const PlayerClient* pSpawn);
inline float PercentHealth(const PlayerClient* pSpawn);
inline float PercentEndurance(const PlayerClient* pSpawn);

static MQPlugin* FindMQ2NavPlugin();

void CastDetrimentalSpells();
void CheckAlias();
void ClearTarget();
void DiscSetup();
void DoINIThings();
void FarmCommand(PlayerClient* pChar, char* Line);
void IgnoreThisCommand(PlayerClient* pChar, char* szLine);
void IgnoreTheseCommand(PlayerClient* pChar, char* szLine);
void ListCommands();
void NavCommand(char* szLine);
void NavEnd();
void NavigateToID(unsigned long ID);
void PermIgnoreCommand(PlayerClient* pchar, char* szline);
void PluginOn();
void PluginOff();
void RestRoutines();
void SummonThings(std::vector<PSPELL>);
void ShowSettings();
void UpdateSearchString();
void UseDiscs();
void VerifyINI(char* ININame, char* Section, char* Key, char* Default);

//End Prototypes

NavCommands.h
#pragma once

using fIsNavMeshLoaded = bool(*)();
using fIsNavPathActive = bool(*)();
using fIsNavPossible = bool(*)(const char* szLine);
using fGetNavPathLength = float(*)(const char* szLine);
using fExecuteNavCommand = bool(*)(const char* szLine);
using fIsNavPathPaused = bool(*)();

unsigned long long NavTimer = 0;

static MQPlugin* FindMQ2NavPlugin()
{
	return GetPlugin("MQ2Nav");
}

float PathLength(unsigned long SpawnID)
{
	MQPlugin* pNav;
	fGetNavPathLength NavPathLength = 0;
	if (pNav = FindMQ2NavPlugin())
	{
		NavPathLength = (fGetNavPathLength)GetProcAddress(pNav->hModule, "GetNavPathLength");
		if (NavPathLength)
		{
			char szNavLine[32] = { 0 };
			sprintf_s(szNavLine, 32, "id %u", SpawnID);
			bool bTemp = gFilterMQ;
			gFilterMQ = true;
			float bTemp2 = NavPathLength(szNavLine);
			gFilterMQ = bTemp;
			return bTemp2;
		}
	}
	return 9999.0f;
}

bool PathExists(unsigned long SpawnID)
{
	MQPlugin* pNav;
	fIsNavPossible NavPathExists = 0;
	if (pNav = FindMQ2NavPlugin())
	{
		NavPathExists = (fIsNavPossible)GetProcAddress(pNav->hModule, "IsNavPossible");
		if (NavPathExists)
		{
			char szNavLine[32] = { 0 };
			sprintf_s(szNavLine, 32, "id %u", SpawnID);
			bool bTemp = gFilterMQ;
			gFilterMQ = true;
			bool bTemp2 = NavPathExists(szNavLine);
			gFilterMQ = bTemp;
			if (bTemp2) return bTemp2;
		}
	}
	return false;
}

bool MeshLoaded()
{
	MQPlugin* pNav;
	fIsNavMeshLoaded NavMeshLoaded = 0;
	if (pNav = FindMQ2NavPlugin())
	{
		//Custom MeshCheck
		NavMeshLoaded = (fIsNavMeshLoaded)GetProcAddress(pNav->hModule, "IsNavMeshLoaded");
		if (NavMeshLoaded)
		{
			return NavMeshLoaded();
		}
	}
	return false;
}

bool NavActive()
{
	MQPlugin* pNav;
	fIsNavPathActive NavPathActive = 0;
	if (pNav = FindMQ2NavPlugin())
	{
		NavPathActive = (fIsNavMeshLoaded)GetProcAddress(pNav->hModule, "IsNavPathActive");
		if (NavPathActive)
		{
			return NavPathActive();
		}
	}
	return false;
}

bool NavPaused()
{
	MQPlugin* pNav;
	fIsNavPathPaused NavPathPaused = 0;
	if (pNav = FindMQ2NavPlugin())
	{
		NavPathPaused = (fIsNavPathPaused)GetProcAddress(pNav->hModule, "IsNavPathPaused");
		if (NavPathPaused)
		{
			return NavPathPaused();
		}
	}
	return false;
}

void NavCommand(char* szLine)
{
	MQPlugin* pNav;
	fExecuteNavCommand TheNavCommand = 0;
	if (pNav = FindMQ2NavPlugin())
	{
		TheNavCommand = (fExecuteNavCommand)GetProcAddress(pNav->hModule, "ExecuteNavCommand");
		if (TheNavCommand)
		{
			char szNavLine[MAX_STRING] = { 0 };
			sprintf_s(szNavLine, MAX_STRING, "%s", szLine);
			bool bTemp = gFilterMQ;
			gFilterMQ = true;
			TheNavCommand(szNavLine);
			gFilterMQ = bTemp;
			NavTimer = GetTickCount64();
		}
	}
	else {
		WriteChatf("Couldn't find MQ2Nav Plugin in the plugin list");
	}
	return;
}

void NavEnd()
{
	MQPlugin* pNav;
	fExecuteNavCommand TheNavCommand = 0;
	if (pNav = FindMQ2NavPlugin())
	{
		TheNavCommand = (fExecuteNavCommand)GetProcAddress(pNav->hModule, "ExecuteNavCommand");
		if (TheNavCommand)
		{
			bool bTemp = gFilterMQ;
			gFilterMQ = true;
			TheNavCommand("stop");
			gFilterMQ = bTemp;
		}
	}
}


ISXEQFarmTest.h
#pragma once
#pragma pack(push)
#pragma pack(8)
#include <isxdk.h>
#pragma pack(pop)


class ISXEQFarmTest :
	public ISXInterface
{
public:

	virtual bool Initialize(ISInterface *p_ISInterface);
	virtual void Shutdown();

	void LoadSettings();
	void ConnectServices();
	void RegisterCommands();
	void RegisterAliases();
	void RegisterDataTypes();
	void RegisterTopLevelObjects();
	void RegisterServices();

	void DisconnectServices();
	void UnRegisterCommands();
	void UnRegisterAliases();
	void UnRegisterDataTypes();
	void UnRegisterTopLevelObjects();
	void UnRegisterServices();

};

extern ISInterface *pISInterface;
extern HISXSERVICE hPulseService;
extern HISXSERVICE hMemoryService;
extern HISXSERVICE hServicesService;

extern HISXSERVICE hEQChatService;
extern HISXSERVICE hEQUIService;
extern HISXSERVICE hEQGamestateService;
extern HISXSERVICE hEQSpawnService;
extern HISXSERVICE hEQZoneService;

extern ISXEQFarmTest *pExtension;
#define printf pISInterface->Printf

#define EzDetour(Address, Detour, Trampoline) IS_Detour(pExtension,pISInterface,hMemoryService,(unsigned int)Address,Detour,Trampoline)
#define EzUnDetour(Address) IS_UnDetour(pExtension,pISInterface,hMemoryService,(unsigned int)Address)

#define EzModify(Address,NewData,Length,Reverse) Memory_Modify(pExtension,pISInterface,hMemoryService,(unsigned int)Address,NewData,Length,Reverse)
#define EzUnModify(Address) Memory_UnModify(pExtension,pISInterface,hMemoryService,(unsigned int)Address)

#define EzHttpRequest(_URL_,_pData_) IS_HttpRequest(pExtension,pISInterface,hHTTPService,_URL_,_pData_)

extern LSType *pStringType;
extern LSType *pIntType;
extern LSType *pBoolType;
extern LSType *pFloatType;
extern LSType *pTimeType;
extern LSType *pByteType;


ISXEQFarmTest.cpp
//
// ISXEQFarmTest
//
#pragma warning(disable:4996)
#include "../ISXEQClient.h"
#include "ISXEQFarmTest.h"

// The mandatory pre-setup function.  Our name is "ISXEQFarmTest", and the class is ISXEQFarmTest.
// This sets up a "ModulePath" variable which contains the path to this module in case we want it,
// and a "PluginLog" variable, which contains the path and filename of what we should use for our
// debug logging if we need it.  It also sets up a variable "pExtension" which is the pointer to
// our instanced class.
ISXPreSetup("ISXEQFarmTest", ISXEQFarmTest);

// Basic LavishScript datatypes, these get retrieved on startup by our initialize function, so we can
// use them in our Top-Level Objects or custom datatypes
LSType *pStringType = 0;
LSType *pIntType = 0;
LSType *pBoolType = 0;
LSType *pFloatType = 0;
LSType *pTimeType = 0;
LSType *pByteType = 0;

ISInterface *pISInterface = 0;
HISXSERVICE hPulseService = 0;
HISXSERVICE hMemoryService = 0;
HISXSERVICE hServicesService = 0;

HISXSERVICE hEQChatService = 0;
HISXSERVICE hEQUIService = 0;
HISXSERVICE hEQGamestateService = 0;
HISXSERVICE hEQSpawnService = 0;
HISXSERVICE hEQZoneService = 0;

unsigned int ISXEQFarmTestXML = 0;

// Forward declarations of callbacks
void __cdecl PulseService(bool Broadcast, unsigned int MSG, void *lpData);
void __cdecl MemoryService(bool Broadcast, unsigned int MSG, void *lpData);
void __cdecl ServicesService(bool Broadcast, unsigned int MSG, void *lpData);


// Initialize is called by Inner Space when the extension should initialize.
bool ISXEQFarmTest::Initialize(ISInterface *p_ISInterface)
{
	pISInterface = p_ISInterface;

	// retrieve basic ISData types
	pStringType = pISInterface->FindLSType("string");
	pIntType = pISInterface->FindLSType("int");
	pBoolType = pISInterface->FindLSType("bool");
	pFloatType = pISInterface->FindLSType("float");
	pTimeType = pISInterface->FindLSType("time");
	pByteType = pISInterface->FindLSType("byte");


	pISInterface->OpenSettings(XMLFileName);
	LoadSettings();

	ConnectServices();

	RegisterCommands();
	RegisterAliases();
	RegisterDataTypes();
	RegisterTopLevelObjects();
	RegisterServices();

	WriteChatf("ISXEQFarmTest Loaded");
	return true;
}

// shutdown sequence
void ISXEQFarmTest::Shutdown()
{
	// save settings, if you changed them and want to save them now.  You should normally save
	// changes immediately.
	//pISInterface->ExportSet(ISXEQFarmTestXML,XMLFileName);

	pISInterface->UnloadSet(ISXEQFarmTestXML);

	DisconnectServices();

	UnRegisterServices();
	UnRegisterTopLevelObjects();
	UnRegisterDataTypes();
	UnRegisterAliases();
	UnRegisterCommands();
}


void ISXEQFarmTest::ConnectServices()
{
	// connect to any services.  Here we connect to "Pulse" which receives a
	// message every frame (after the frame is displayed) and "Memory" which
	// wraps "detours" and memory modifications
	hPulseService = pISInterface->ConnectService(this, "Pulse", PulseService);
	hMemoryService = pISInterface->ConnectService(this, "Memory", MemoryService);
	hServicesService = pISInterface->ConnectService(this, "Services", ServicesService);


}

void ISXEQFarmTest::RegisterCommands()
{
	// add any commands
	//    pISInterface->AddCommand("MyCommand",MyCommand,true,false);

}

void ISXEQFarmTest::RegisterAliases()
{
	// add any aliases
}

void ISXEQFarmTest::RegisterDataTypes()
{
	// add any datatypes
	// pMyType = new MyType;
	// pISInterface->AddLSType(*pMyType);
}

void ISXEQFarmTest::RegisterTopLevelObjects()
{
	// add any Top-Level Objects
	//    pISInterface->AddTopLevelObject("MapSpawn",tloMapSpawn);
}

void ISXEQFarmTest::RegisterServices()
{
	// register any services.  Here we demonstrate a service that does not use a
	// callback
	// set up a 1-way service (broadcast only)
	//    hISXEQFarmTestService=pISInterface->RegisterService(this,"ISXEQFarmTest Service",0);
	// broadcast a message, which is worthless at this point because nobody will receive it
	// (nobody has had a chance to connect)
	//    pISInterface->ServiceBroadcast(this,hISXEQFarmTestService,ISXSERVICE_MSG+1,0);

}

void ISXEQFarmTest::DisconnectServices()
{
	// gracefully disconnect from services
	if (hPulseService)
		pISInterface->DisconnectService(this, hPulseService);
	if (hMemoryService)
	{
		pISInterface->DisconnectService(this, hMemoryService);
		// memory modifications are automatically undone when disconnecting
		// also, since this service accepts messages from clients we should reset our handle to
		// 0 to make sure we dont try to continue using it
		hMemoryService = 0;
	}
	if (hServicesService)
		pISInterface->DisconnectService(this, hServicesService);
}

void ISXEQFarmTest::UnRegisterCommands()
{
	// remove commands
	//    pISInterface->RemoveCommand("MyCommand");
}
void ISXEQFarmTest::UnRegisterAliases()
{
	// remove aliases
}
void ISXEQFarmTest::UnRegisterDataTypes()
{
	// remove data types
	//if (pMyType)
	//{
	//    pISInterface->RemoveLSType(*pMyType);
	//    delete pMyType;
	//}
}
void ISXEQFarmTest::UnRegisterTopLevelObjects()
{
	// remove Top-Level Objects
	//    pISInterface->RemoveTopLevelObject("MapSpawn");
}
void ISXEQFarmTest::UnRegisterServices()
{
	// shutdown our own services
	//    if (hISXEQFarmTestService)
	//        pISInterface->ShutdownService(this,hISXEQFarmTestService);
}

void ISXEQFarmTest::LoadSettings()
{
	bool bChanged = false;

	// open the settings file, and store the ID number in ISXEQFarmTestXML
	ISXEQFarmTestXML = pISInterface->OpenSettings(XMLFileName);
	// This ID is used to refer to the base set of the loaded settings

	/*
	Here is an example settings hierarchy:
	base set
	..settings
	.."General" set
	.."Characters" set
	...."Server A" set
	......"Character A" set
	........settings
	......"Character B" set
	........settings
	...."Server B" set
	......"Character A" set
	........settings
	...."Server C" set
	......"Character A" set
	........settings
	......"Character B" set
	........settings
	......"Character C" set
	........settings

	As you can see, there can be any number of nested sets, which can contain sets, settings, and comments.
	Any set can be referred to by its ID.  The base set is stored in ISXEQFarmTestXML.  Note that the settings
	are NOT "xml" after loading them, they are simply settings stored in memory!
	*/
	// Get General set ID, which can be a child of the base set
	if (unsigned int GeneralGUID = pISInterface->FindSet(ISXEQFarmTestXML, "General"))
	{
		// retrieve a text setting from the General set
		char Text[512];
		if (pISInterface->GetSetting(GeneralGUID, "Name of setting", Text, 512))
		{
			// success
		}
		unsigned int IntSetting;
		if (pISInterface->GetSetting(GeneralGUID, "Name of int setting", IntSetting))
		{
			// success
		}
	}
	/*
	Deeper sets and settings are accessed in exactly the same way.  If a sub-set of General was to be used,
	accessing it would work like so:
	if (unsigned int SubSetGUID=pISInterface->FindSet(GeneralGUID,"Sub-set name"))
	{
	// got the sub-set
	}
	*/

	// save settings if we changed them
	if (bChanged)
		pISInterface->ExportSet(ISXEQFarmTestXML, XMLFileName);
	// Note: ExportSet can save ANY set and its settings, sets, and comments as ANY filename.  The "General" set
	// could be stored as "General.XML" simply by doing pISInterface->ExportSet(GeneralGUID,"General.XML");
}


void __cdecl PulseService(bool Broadcast, unsigned int MSG, void *lpData)
{
	if (MSG == PULSE_PULSE)
	{
		// Same as OnPulse
	}
}

void __cdecl MemoryService(bool Broadcast, unsigned int MSG, void *lpData)
{
	// no messages are currently associated with this service (other than
	// system messages such as client disconnect), so do nothing.
}

void __cdecl EQUIService(bool Broadcast, unsigned int MSG, void *lpData)
{
	switch (MSG)
	{
	case UISERVICE_CLEANUP:
		// same as OnCleanUI
		break;
	case UISERVICE_RELOAD:
		// same as OnReloadUI
		break;
	}
}
void __cdecl EQGamestateService(bool Broadcast, unsigned int MSG, void *lpData)
{
#define GameState ((unsigned int)lpData)
	if (MSG == GAMESTATESERVICE_CHANGED)
	{
		// same as SetGameState
	}
#undef GameState
}
void __cdecl EQSpawnService(bool Broadcast, unsigned int MSG, void *lpData)
{
	switch (MSG)
	{
#define pSpawn ((PSPAWNINFO)lpData)
	case SPAWNSERVICE_ADDSPAWN:
		// same as OnAddSpawn
		break;
	case SPAWNSERVICE_REMOVESPAWN:
		// same as OnRemoveSpawn
		break;
#undef pSpawn
#define pGroundItem ((PGROUNDITEM)lpData)
	case SPAWNSERVICE_ADDITEM:
		// same as OnAddGroundItem
		break;
	case SPAWNSERVICE_REMOVEITEM:
		// same as OnRemoveGroundItem
		break;
#undef pGroundItem
	}
}

void __cdecl EQZoneService(bool Broadcast, unsigned int MSG, void *lpData)
{
	switch (MSG)
	{
	case ZONESERVICE_BEGINZONE:
		// same as OnBeginZone
		break;
	case ZONESERVICE_ENDZONE:
		// same as OnEndZone
		break;
	case ZONESERVICE_ZONED:
		// same as OnZoned
		break;
	}
}

void __cdecl EQChatService(bool Broadcast, unsigned int MSG, void *lpData)
{
#define pChat ((_EQChat*)lpData)
	switch (MSG)
	{
	case CHATSERVICE_OUTGOING:
		// same as OnWriteChatColor
		break;
	case CHATSERVICE_INCOMING:
		// same as OnIncomingChat
		break;
	}
#undef pChat
}

// This uses the Services service to connect to ISXEQ services
void __cdecl ServicesService(bool Broadcast, unsigned int MSG, void *lpData)
{
#define Name ((char*)lpData)
	switch (MSG)
	{
	case SERVICES_ADDED:
		if (!stricmp(Name, "EQ UI Service"))
		{
			hEQUIService = pISInterface->ConnectService(pExtension, Name, EQUIService);
		}
		else if (!stricmp(Name, "EQ Gamestate Service"))
		{
			hEQGamestateService = pISInterface->ConnectService(pExtension, Name, EQGamestateService);
		}
		else if (!stricmp(Name, "EQ Spawn Service"))
		{
			hEQSpawnService = pISInterface->ConnectService(pExtension, Name, EQSpawnService);
		}
		else if (!stricmp(Name, "EQ Zone Service"))
		{
			hEQZoneService = pISInterface->ConnectService(pExtension, Name, EQZoneService);
		}
		else if (!stricmp(Name, "EQ Chat Service"))
		{
			hEQChatService = pISInterface->ConnectService(pExtension, Name, EQChatService);
		}
		break;
	case SERVICES_REMOVED:
		if (!stricmp(Name, "EQ UI Service"))
		{
			if (hEQUIService)
			{
				pISInterface->DisconnectService(pExtension, hEQUIService);
				hEQUIService = 0;
			}
		}
		else if (!stricmp(Name, "EQ Gamestate Service"))
		{
			if (hEQGamestateService)
			{
				pISInterface->DisconnectService(pExtension, hEQGamestateService);
				hEQGamestateService = 0;
			}
		}
		else if (!stricmp(Name, "EQ Spawn Service"))
		{
			if (hEQSpawnService)
			{
				pISInterface->DisconnectService(pExtension, hEQSpawnService);
				hEQSpawnService = 0;
			}
		}
		else if (!stricmp(Name, "EQ Zone Service"))
		{
			if (hEQZoneService)
			{
				pISInterface->DisconnectService(pExtension, hEQZoneService);
				hEQZoneService = 0;
			}
		}
		else if (!stricmp(Name, "EQ Chat Service"))
		{
			if (hEQChatService)
			{
				pISInterface->DisconnectService(pExtension, hEQChatService);
				hEQChatService = 0;
			}
		}
		break;
	}
#undef Name
}

int MyCommand(int argc, char *argv[])
{
	// IS Commands work just like console programs. argc is the argument (aka parameter) count, 
	// argv is an array containing them.  The name of the command is always the first argument.
	if (argc<2)
	{
		// This command requires arguments
		WriteChatf("Syntax: %s <text>", argv[0]);
		return 0;
	}

	char FullText[8192] = { 0 };
	pISInterface->GetArgs(1, argc, argv, FullText, sizeof(FullText)); // this gets the rest of the arguments, from 1 to argc
	WriteChatf("You passed the following text to the command:");
	WriteChatf("%s", FullText);

	// For most commands, you need to deal with only one argument at a time, instead of the entire line.
	for (int i = 1; i < argc; i++)
	{
		WriteChatf("argv[%d]=\"%s\"", i, argv[i]);
	}

	// return 0 or greater for normal conditions, or return -1 for an error that should stop
	// a script. 
	return 0;
}


NavCommands.cpp
#include "NavCommands.h"

static PMQPLUGIN FindMQ2NavPlugin()
{
	PMQPLUGIN pPlugin = pPlugins;
	while (pPlugin)
	{
		if (!_stricmp("MQ2Nav", pPlugin->szFilename))
		{
			return pPlugin;
		}

		pPlugin = pPlugin->pNext;
	}

	return nullptr;
}

float PathLength(DWORD SpawnID)
{
	PMQPLUGIN pNav;
	fNavPathLength NavPathLength = 0;
	if (pNav = FindMQ2NavPlugin())
	{
		NavPathLength = (fNavPathLength)GetProcAddress(pNav->hModule, "NavPathLength");
		if (NavPathLength)
		{
			CHAR szNavLine[32];
			sprintf_s(szNavLine, 32, "id %u", SpawnID);
			bool bTemp = gFilterMQ;
			gFilterMQ = true;
			float bTemp2 = NavPathLength(szNavLine);
			gFilterMQ = bTemp;
			return bTemp2;
		}
	}
	return 9999.0f;
}
bool PathExists(DWORD SpawnID)
{
	PMQPLUGIN pNav;
	fNavPathExists NavPathExists = 0;
	if (pNav = FindMQ2NavPlugin())
	{
		NavPathExists = (fNavPathExists)GetProcAddress(pNav->hModule, "NavPossible");
		if (NavPathExists)
		{
			CHAR szNavLine[32];
			sprintf_s(szNavLine, 32, "id %u", SpawnID);
			bool bTemp = gFilterMQ;
			gFilterMQ = true;
			bool bTemp2 = NavPathExists(szNavLine);
			WriteChatf("Temp2: %d", bTemp2);
			gFilterMQ = bTemp;
			return bTemp2;
		}
	}
	return false;
}

bool MeshLoaded()
{
	PMQPLUGIN pNav;
	fNavMeshLoaded NavMeshLoaded = 0;
	if (pNav = FindMQ2NavPlugin())
	{
		//Custom MeshCheck
		NavMeshLoaded = (fNavMeshLoaded)GetProcAddress(pNav->hModule, "NavMeshLoaded"); //my MeshCheck
		if (!NavMeshLoaded)
			NavMeshLoaded = (fNavMeshLoaded)GetProcAddress(pNav->hModule, "IsMeshLoaded"); //eqmule MeshCheck
		if (NavMeshLoaded)
		{
			return NavMeshLoaded();
		}
	}
	return false;
}

bool NavActive()
{
	PMQPLUGIN pNav;
	fNavPathActive NavPathActive = 0;
	if (pNav = FindMQ2NavPlugin())
	{
		NavPathActive = (fNavMeshLoaded)GetProcAddress(pNav->hModule, "NavPathActive");
		if (NavPathActive)
		{
			return NavPathActive();
		}
	}
	return false;
}

bool NavPaused()
{
	PMQPLUGIN pNav;
	fNavPathPaused NavPathPaused = 0;
	if (pNav = FindMQ2NavPlugin())
	{
		NavPathPaused = (fNavPathPaused)GetProcAddress(pNav->hModule, "NavPathPaused");
		if (NavPathPaused)
		{
			return NavPathPaused();
		}
	}
	return false;
}

VOID NavCommand(PSPAWNINFO pChar, PCHAR szLine)
{
	PMQPLUGIN pNav;
	fNavCommand TheNavCommand = 0;
	if (pNav = FindMQ2NavPlugin())
	{
		TheNavCommand = (fNavCommand)GetProcAddress(pNav->hModule, "NavCommand");
		if (TheNavCommand)
		{
			CHAR szNavLine[32];
			sprintf_s(szNavLine, 32, "id %s", szLine);
			bool bTemp = gFilterMQ;
			gFilterMQ = true;
			TheNavCommand(pChar, szNavLine);
			gFilterMQ = bTemp;
		}

	}
	return;
}

void NavEnd(PSPAWNINFO pChar)
{
	PMQPLUGIN pNav;
	fNavCommand TheNavCommand = 0;
	if (pNav = FindMQ2NavPlugin())
	{
		TheNavCommand = (fNavCommand)GetProcAddress(pNav->hModule, "NavCommand");
		if (TheNavCommand)
		{
			bool bTemp = gFilterMQ;
			gFilterMQ = true;
			TheNavCommand(pChar, "stop");
			gFilterMQ = bTemp;
		}
	}
}

void NavigateToID(DWORD ID) {
	PSPAWNINFO Mob = (PSPAWNINFO)GetSpawnByID(ID);
	CHAR szNavInfo[32];
	sprintf_s(szNavInfo, 32, "%u", ID);
	if (Mob && (!LineOfSight(GetCharInfo()->pSpawn, Mob) || GetDistance(GetCharInfo()->pSpawn, Mob) > 15)) {
		if (!NavActive()) {
			NavCommand(GetCharInfo()->pSpawn, szNavInfo);

		}
	}
	if (LineOfSight(GetCharInfo()->pSpawn, Mob) && GetDistance(GetCharInfo()->pSpawn, Mob) < 15) {
		if (NavActive()) NavEnd(GetCharInfo()->pSpawn);
		if (Mob && pTarget && ((PSPAWNINFO)pTarget)->SpawnID != Mob->SpawnID || !pTarget) {
			TargetIt(Mob);
		}
		if (GetCharInfo()->pSpawn->StandState == STANDSTATE_SIT)
			EzCommand("/stand");
		//Doesn't appear to work, always returns -0 if GetCharInfo or 0 if Mob
		//WriteChatf("BearingToTarget: %f", GetCharInfo()->pSpawn->BearingToTarget);
		if (fabs(Mob->Heading - GetCharInfo()->pSpawn->Heading) < 240 || fabs(Mob->Heading - GetCharInfo()->pSpawn->Heading) > 265) Face(GetCharInfo()->pSpawn, "fast");

	}
	if (Mob && pTarget) {
		if (CastDetrimental) CastDetrimentalSpells();
		if (*EQADDR_ATTACK) return;
		*EQADDR_ATTACK = 00000001;
		EzCommand("/pet attack");
		EzCommand("/pet swarm");

	}
}